### PR TITLE
Fix specificity issues when using the `@headlessui/tailwindcss` package

### PR DESCRIPTION
**Pull request by RobinMalfait**
This PR fixes specifity related issues when using the `@headlessui/tailwindcss` on the `Listbox` component.

```js
<Listbox value={selected} onChange={setSelected}>
  <Listbox.Button>Open Listbox</Listbox.Button>
  <Listbox.Options>
    {people.map((person, personIdx) => (
      <Listbox.Option
        key={personIdx}
        className="ui-active:bg-amber-100 ui-active:text-amber-900 ui-not-active:text-gray-900"
        value={person}
      >
        <span className="ui-selected:font-medium ui-not-selected:font-normal">
          {person.name}
        </span>

        <span className="ui-selected:flex ui-not-selected:hidden">
          <CheckIcon className="h-5 w-5" aria-hidden="true" />
        </span>
      </Listbox.Option>
    ))}
  </Listbox.Options>
</Listbox>
```

- When the `Listbox` is open, then
  - The `Listbox.Options` will have the following data attribute: `data-headlessui-state="open"`
  - The selected `Listbox.Option` will have the followin data attribute: `data-headlessui-state="active selected"`
  - The span for the icon in the `Listbox.Option` has the following classes: `class="ui-selected:flex ui-not-selected:hidden"`

You would expect that the `ui-selected` is matching because the `Listbox.Option` _is_ the selected one. However, the `ui-not-selected` is _also_ matching. The reason for this is because the selector is defined as:

```css
:where([data-headlessui-state]:not([data-headlessui-state~="selected"]))
  .ui-not-selected\:hidden:not([data-headlessui-state]) {
  display: none;
}
```

The reason for this is beacuse we wanted to style an element when a given state is present, or when a given state is _not_ present. Looking at the `Listbox.Options` it _does_ match the `[data-headlessui-state]` and it also matches the `:not([data-headlessui-state~="selected"])` because it only has the `open` state, therefore the css selector matches.

This PR tries to solve that by being very explicit. Sadly this does increase the HTML a bit, but this is the solution:

- When the `Listbox` is open, then
  - The `Listbox.Options` will have the following data attribute: `data-headlessui-state="open"`
  - The selected `Listbox.Option` will have the followin data attribute: `data-headlessui-state="active selected not-disabled"`
  - The span for the icon in the `Listbox.Option` has the following classes: `class="ui-selected:flex ui-not-selected:hidden"`

It's very similar, but instead of only having the states that apply, we also expose the states that _don't_ apply. For example by using `not-disabled`.

The `Listbox.Options` only has an open state so it doesn't contain other states right now.

This on its own is not enough, we also have to modify the css selector. We simplified it to:

```css
:where([data-headlessui-state~="not-selected"]) .ui-not-selected\:hidden {
  display: none;
}
```

In short, this is the diff we applied:

```diff
- `&[data-headlessui-state]:not([data-headlessui-state~="${state}"])`
- `:where([data-headlessui-state]:not([data-headlessui-state~="${state}"])) &:not([data-headlessui-state])`,
+ `&[data-headlessui-state~="not-${state}"]`,
+ `:where([data-headlessui-state~="not-${state}"]) &`,
```

Reproduction of the bug in CodeSandbox: https://codesandbox.io/s/agitated-fermi-8uqhf7?file=/src/App.js
There should be a `checkmark` icon for the selected person, but there is none.

Fixes: #2291

### CODE REVIEW

**Comment by RobinMalfait** [comment]
Closing for now. We want to rethink this by scoping the states to components itself. Right now we expose `ui-open` for example, but if you have a `Listbox` inside a `Disclosure` then there is no way to tell which `open` to use (both `Listbox` and `Disclosure` expose them). The system scoped to components should improve on this.
